engine/src/main/java/io/zeebe/engine/processing/streamprocessor/StreamProcessor.java

```diff
@@ -312,6 +315,18 @@ private void onFailure(final Throwable throwable) {
     }
   }
 
+  private void onUnrecoverableFailure(final Exception e) {
```

Comment:
Move this into `onFailure()` and handle it as a special case. Most of the logic is shared. Besides, `onFailure` is called any time there is an uncaught exception.